File: .claude/hooks/check_project_lint.py
#!/usr/bin/env python3
"""
Stop hook for Claude Code: comprehensive lint + type-check at end of turn.

Runs the full project ESLint and TypeScript compiler, then returns
any issues as additionalContext so Claude can fix them before
declaring the task complete.
"""

import json
import os
import subprocess
import sys


def run_eslint_full(project_dir: str) -> str | None:
    """Run ESLint on the full project with JSON output."""
    try:
        result = subprocess.run(
            ["npx", "eslint", ".", "-f", "json"],
            cwd=project_dir,
            capture_output=True,
            text=True,
            timeout=120,
        )
    except subprocess.TimeoutExpired:
        return "ESLint: timed out after 120s"
    except FileNotFoundError:
        return None

    if result.returncode not in (0, 1):
        return None

    if not result.stdout.strip():
        return None

    try:
        reports = json.loads(result.stdout)
    except json.JSONDecodeError:
        return None

    total_errors = 0
    total_warnings = 0
    fixable_errors = 0
    fixable_warnings = 0
    file_lines: list[str] = []

    for report in reports:
        file_path = report.get("filePath", "")
        if project_dir and file_path.startswith(project_dir):
            file_path = file_path[len(project_dir) + 1:]

        messages = report.get("messages", [])
        if not messages:
            continue

        file_errors = sum(1 for m in messages if m.get("severity") == 2)
        file_warnings = sum(1 for m in messages if m.get("severity") == 1)
        file_fixable = sum(1 for m in messages if m.get("fix") and m.get("severity") == 2)
        file_fixable_w = sum(1 for m in messages if m.get("fix") and m.get("severity") == 1)

        total_errors += file_errors
        total_warnings += file_warnings
        fixable_errors += file_fixable
        fixable_warnings += file_fixable_w

        file_lines.append(
            f"  `{file_path}` ({file_errors} error{'s' if file_errors != 1 else ''}, "
            f"{file_warnings} warning{'s' if file_warnings != 1 else ''}):"
        )

        for msg in messages:
            line = msg.get("line", "?")
            col = msg.get("column", "?")
            rule = msg.get("ruleId", "unknown")
            severity = "error" if msg.get("severity") == 2 else "warning"
            message_text = msg.get("message", "").split("\n")[0]
            fix_tag = " [auto-fixable]" if msg.get("fix") else ""
            file_lines.append(f"    L{line}:{col}  {severity}  {rule}  {message_text}{fix_tag}")

    if total_errors == 0 and total_warnings == 0:
        return None

    summary = (
        f"ESLint: {total_errors} error{'s' if total_errors != 1 else ''}, "
        f"{total_warnings} warning{'s' if total_warnings != 1 else ''}"
    )
    if fixable_errors > 0 or fixable_warnings > 0:
        summary += (
            f" ({fixable_errors} fixable error{'s' if fixable_errors != 1 else ''}, "
            f"{fixable_warnings} fixable warning{'s' if fixable_warnings != 1 else ''})"
        )

    return summary + "\n" + "\n".join(file_lines)


def run_type_check(project_dir: str) -> str | None:
    """Run TypeScript compiler, return formatted output or None if clean."""
    try:
        result = subprocess.run(
            ["npx", "tsc", "--noEmit"],
            cwd=project_dir,
            capture_output=True,
            text=True,
            timeout=120,
        )
    except subprocess.TimeoutExpired:
        return "TypeScript: timed out after 120s"
    except FileNotFoundError:
        return None

    if result.returncode == 0:
        return None

    output = (result.stdout or "") + (result.stderr or "")
    output = output.strip()
    if not output:
        return None

    error_count = output.count("error TS")
    lines = output.split("\n")
    if len(lines) > 30:
        output = "\n".join(lines[:30]) + f"\n  ... and {len(lines) - 30} more line(s)"

    return f"TypeScript ({error_count} error{'s' if error_count != 1 else ''}):\n{output}"


def main() -> None:
    project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())

    sections: list[str] = []

    eslint_output = run_eslint_full(project_dir)
    if eslint_output:
        sections.append(eslint_output)

    tsc_output = run_type_check(project_dir)
    if tsc_output:
        sections.append(tsc_output)

    if not sections:
        sys.exit(0)

    combined = "\n\n".join(sections)

    output = {
        "hookSpecificOutput": {
            "hookEventName": "Stop",
            "additionalContext": (
                "Full project lint and type-check results. "
                "Please review and fix the issues below:\n\n"
                f"{combined}"
            ),
        }
    }

    json.dump(output, sys.stdout)
    sys.stdout.flush()


if __name__ == "__main__":
    main()

File: .claude/hooks/check_project_lint.mjs
import { spawnSync } from 'node:child_process';

function runEslintFull(projectDir) {
  const result = spawnSync('npx', ['eslint', '.', '-f', 'json'], {
    cwd: projectDir,
    encoding: 'utf8',
  });

  if (result.status !== 0 && result.status !== 1) return null;
  if (!result.stdout?.trim()) return null;

  let reports;
  try {
    reports = JSON.parse(result.stdout);
  } catch {
    return null;
  }

  let totalErrors = 0;
  let totalWarnings = 0;
  let fixableErrors = 0;
  let fixableWarnings = 0;
  const fileLines = [];

  for (const report of reports) {
    let filePath = report?.filePath || '';
    if (projectDir && filePath.startsWith(projectDir)) {
      filePath = filePath.slice(projectDir.length + 1);
    }

    const messages = Array.isArray(report?.messages) ? report.messages : [];
    if (messages.length === 0) continue;

    const fileErrors = messages.filter((m) => m?.severity === 2).length;
    const fileWarnings = messages.filter((m) => m?.severity === 1).length;
    const fileFixable = messages.filter((m) => m?.severity === 2 && m?.fix).length;
    const fileFixableWarnings = messages.filter((m) => m?.severity === 1 && m?.fix).length;

    totalErrors += fileErrors;
    totalWarnings += fileWarnings;
    fixableErrors += fileFixable;
    fixableWarnings += fileFixableWarnings;

    fileLines.push(`  \`${filePath}\` (${fileErrors} error${fileErrors !== 1 ? 's' : ''}, ${fileWarnings} warning${fileWarnings !== 1 ? 's' : ''}):`);

    for (const msg of messages) {
      const line = msg?.line ?? '?';
      const col = msg?.column ?? '?';
      const rule = msg?.ruleId ?? 'unknown';
      const severity = msg?.severity === 2 ? 'error' : 'warning';
      const messageText = String(msg?.message ?? '').split('\n')[0];
      const fixTag = msg?.fix ? ' [auto-fixable]' : '';
      fileLines.push(`    L${line}:${col}  ${severity}  ${rule}  ${messageText}${fixTag}`);
    }
  }

  if (totalErrors === 0 && totalWarnings === 0) return null;

  let summary = `ESLint: ${totalErrors} error${totalErrors !== 1 ? 's' : ''}, ${totalWarnings} warning${totalWarnings !== 1 ? 's' : ''}`;
  if (fixableErrors > 0 || fixableWarnings > 0) {
    summary += ` (${fixableErrors} fixable error${fixableErrors !== 1 ? 's' : ''}, ${fixableWarnings} fixable warning${fixableWarnings !== 1 ? 's' : ''})`;
  }

  return `${summary}\n${fileLines.join('\n')}`;
}

function runTypeCheck(projectDir) {
  const result = spawnSync('npx', ['tsc', '--noEmit'], {
    cwd: projectDir,
    encoding: 'utf8',
  });

  if (result.status === 0) return null;

  const output = `${result.stdout || ''}${result.stderr || ''}`.trim();
  if (!output) return null;

  const errorCount = (output.match(/error TS/g) || []).length;
  const lines = output.split('\n');
  const truncated = lines.length > 30 ? `${lines.slice(0, 30).join('\n')}\n  ... and ${lines.length - 30} more line(s)` : output;

  return `TypeScript (${errorCount} error${errorCount !== 1 ? 's' : ''}):\n${truncated}`;
}

const projectDir = process.env.CLAUDE_PROJECT_DIR || process.cwd();
const sections = [];

const eslintOutput = runEslintFull(projectDir);
if (eslintOutput) sections.push(eslintOutput);

const tscOutput = runTypeCheck(projectDir);
if (tscOutput) sections.push(tscOutput);

if (!sections.length) process.exit(0);

process.stdout.write(JSON.stringify({
  hookSpecificOutput: {
    hookEventName: 'Stop',
    additionalContext: `Full project lint and type-check results. Please review and fix the issues below:\n\n${sections.join('\n\n')}`,
  },
}));
